DEFI/LiquidityPool/certora/specs/Full.spec

```diff
@@ -0,0 +1,448 @@
+
+    uint256 sharesA = deposit(e, amountA) at init;
+    uint256 sharesB = deposit(e, amountB);
+
+    assert to_mathint(wholeShares) >= sharesA + sharesB;
+}
+
+/*
+┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
+│ Withdraw                                                                                                         │
+└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+*/
+rule withdrawIntegrity(env e){
+
+    require e.msg.sender != currentContract; // this assumption must hold to avoid shares dilute attack
+
+    uint256 amount;
+    uint256 clientBalanceBefore = underlying.balanceOf(e.msg.sender);
+    uint256 clientSharesBefore = balanceOf(e.msg.sender);
+    
+    withdraw(e, amount);
+
+    uint256 clientBalanceAfter = underlying.balanceOf(e.msg.sender);
+    uint256 clientSharesAfter = balanceOf(e.msg.sender);
+
+    assert (amount == 0) => (clientBalanceBefore == clientBalanceAfter) && (clientSharesBefore == clientSharesAfter);
+    assert (amount > 0) => (clientBalanceBefore < clientBalanceAfter) && (clientSharesBefore > clientSharesAfter);
+}
+
+rule withdrawRevertConditions(env e){
+
+    require e.msg.sender != currentContract; // this assumption must hold to avoid shares dilute attack
+
+    uint256 amount;
+    uint256 clientBalanceBefore = underlying.balanceOf(e.msg.sender);
+    uint256 clientSharesBefore = balanceOf(e.msg.sender);
+    
+    bool clientBalanceUnderflow = clientSharesBefore - amount < 0;
+    bool poolUnderflow = underlying.balanceOf(currentContract) - sharesToAmount(amount) < 0 || totalSupply() - amount < 0;
+    bool underflow = clientBalanceUnderflow || poolUnderflow;
+    bool overflow = clientBalanceBefore + sharesToAmount(amount) > max_uint256;
+    bool payable = e.msg.value != 0;
+    bool reentrancy = ghostReentrancyStatus == 2;
+    bool notEnoughAllowance = underlying.allowance(currentContract, currentContract) < sharesToAmount(amount);
+    bool zeroAmount = sharesToAmount(amount) == 0;
+    bool poolIsEmpty = underlying.balanceOf(currentContract) == 0;
+    bool expectedRevert = poolIsEmpty || underflow || overflow || payable || reentrancy || notEnoughAllowance || zeroAmount;
+
+    withdraw@withrevert(e, amount);
+
+    assert lastReverted <=> expectedRevert;
+}
+
+rule withdrawDontAffectThirdParty(env e){
+    address thirdParty;
+    uint256 amount;
+
+    require thirdParty != e.msg.sender && thirdParty != currentContract;
+
+    uint256 thirdPartyBalanceBefore = underlying.balanceOf(thirdParty);
+    uint256 thirdPartySharesBefore = balanceOf(thirdParty);
+
+    withdraw(e, amount);
+
+    uint256 thirdPartyBalanceAfter = underlying.balanceOf(thirdParty);
+    uint256 thirdPartySharesAfter = balanceOf(thirdParty);
+
+    assert thirdPartyBalanceBefore == thirdPartyBalanceAfter;
+    assert thirdPartySharesBefore == thirdPartySharesAfter;
```

Comment:
can we have this as a parametric rule?

Reply:
maybe it does not hold for flashloan, but still I prefer parametric rules instead of copy paste

Reply:
Done